torchprime/sharding/shard_model.py

```diff
@@ -0,0 +1,173 @@
+from collections.abc import Callable
+
+import torch.nn
+
+ShardWeightFn = Callable[[torch.Tensor, str], torch.Tensor]
+"""
+ShardWeightFn optionally transforms a weight tensor based on its name.
+
+Args:
+
+  weight (torch.Tensor): The weight tensor to be transformed.
+  
+  name (str): The name of the weight tensor as it appears in the state dict.
+
+Returns:
+
+  torch.Tensor: The transformed weight tensor.
+
+"""
+
+
+ShardActivationFn = Callable[[torch.nn.Module, str], torch.nn.Module]
+"""
+ShardActivationFn optionally transforms a module based on its name.
+
+Args:
+
+  module (torch.nn.Module): The module to be transformed.
+  
+  name (str): The name of the module as it appears in the state dict.
+
+Returns:
+
+  torch.nn.Module: The transformed module, or the original module if
+    no transformation is needed.
+
+"""
+
+
+def shard_model(
+  model: torch.nn.Module,
+  shard_weight: ShardWeightFn,
+  shard_activation: ShardActivationFn,
+) -> torch.nn.Module:
+  """
+  Transforms `model` by applying `shard_weight` to each weight tensor and applying
+  `shard_activation` to each module. Returns the transformed module.
+  """
+  state_dict = {}
+  for name, param in model.state_dict().items():
+    state_dict[name] = shard_weight(param, name)
+  model.load_state_dict(state_dict, assign=True)
+  return _wrap_module(model, shard_activation, tuple())
+
+
+def _wrap_module(
+  mod: torch.nn.Module, shard_activation: ShardActivationFn, prefix: tuple[str, ...]
+) -> torch.nn.Module:
+  """
+  Recursively transforms the modules to shard activations.
+
+  Start from the leaf modules and work our way up, to handle cases where one
+  module is the child of another. The child modules will be transformed first,
+  and then the parent module will be transformed, possibly with transformed
+  children.
+  """
+  new_children = {}
+  for name, child in mod.named_children():
+    new_children[name] = _wrap_module(child, shard_activation, prefix + (name,))
+  for name, new_child in new_children.items():
+    mod.set_submodule(name, new_child)
+  return shard_activation(mod, ".".join(prefix))
+
+
+def shard_model_from_config(
```

Comment:
Do you think it'll be easier if we just accept a mesh argument and pass the mesh argument to shard_output and shard_param callables?

Reply:
That would be convenient but I wanted this to work across `torchax` and `torch_xla`, which takes different meshes. I think it's still useful to have a convenience wrapper around this, so I'll add a TODO here to address in the next PR.

Reply:
Filed https://github.com/AI-Hypercomputer/torchprime/issues/86